### ISSUE TITLE
Possible memory leak in ProgressTracker

### ISSUE THREAD

**Issue by edwardmp** [bustoutsolutions/siesta#268]
Hi,

Just found a memory leak using Instruments in Siesta.
Seems to be related to the timer in ProgressTracker.swift, but not the obvious thing (strong reference from within timer closure).

![Image](https://cl.ly/101G3z2l2Q2V/Schermafbeelding%2525202018-08-16%252520om%25252021.09.59.png)

**Comment by jwells89** [comment]
I'm seeing this as well. I'd have to roll back and verify to make sure, but I believe that this issue appeared after upgrading to 1.4.0.

**Comment by pcantrell** [comment]
Thanks for flagging this. 1.4.0 had some progress-related refactoring, which seems a like culprit. I'll poke at it tomorrow or Wed unless somebody else gets to it first.

**Comment by pcantrell** [comment]
I found a likely culprit. Could you check if the `memory-leak` branch fixes the problem for you?

**Comment by jwells89** [comment]
I just tested `memory-leak` branch on a couple of different projects and it looks like it does indeed fix the problem.

**Comment by pcantrell** [comment]
Excellent! Thanks. I'll cut a release just for this fix as soon as I can find an hour.

I’d love to figure out how to regression test this by checking for total deallocation of all Siesta classes after the test suite runs, but without stuffing instance-counting code into Siesta itself. Anyone know of any Apple toolchain trickery that might help?

**Comment by edwardmp** [comment]
@pcantrell give me 30 minutes then I'll test it as well. Great job by the way.
Without diving into the code: wouldn't making self weak in a capture list in the `cacheRequest.chained` closure achieve the same and fix this leak?

**Comment by pcantrell** [comment]
No, because the problem was that code in `chained` is not guaranteed to be called (if for example the first request in the chain is cancelled), so cleanup code always belongs in `onCompletion` — a detail I overlooked despite the fact that _my own docs_ make a point of it! 🙄

**Comment by edwardmp** [comment]
@pcantrell Oh right, so the issue here was that that code was not called in some circumstances and that led to the timer not being deallocated down the line?

**Comment by pcantrell** [comment]
Right. The chained request (“try cache first, then network”) that method creates was sticking around in `loadRequest` forever if it was cancelled before it reached the “try the network” part.

**Comment by pcantrell** [comment]
> Do see some other leaks than include Siesta stuff in the stack trace, but for now I'm assuming those are related to something else.

It is possible Siesta has more leaks lurking. I can see from this issue that I need a full audit and an ongoing testing strategy for them.

However, Siesta’s “keep resources in case they’re needed” algorithm does make it easy to get false positives on leaks. `Service` and `Resource` form an intentional retain cycle until one of three things triggers an unused resource cleanup:

- a low memory event on iOS,
- the resource cache exceeds [`cachedResourceCountLimit`](https://bustoutsolutions.github.io/siesta/api/Classes/Service.html#//apple_ref/swift/Property/cachedResourceCountLimit), or
- you manually call [`flushUnusedResources()`](https://bustoutsolutions.github.io/siesta/api/Classes/Service.html#//apple_ref/swift/Method/flushUnusedResources()).

Unless one of those things happens right before you inspect for leaks, you’ll see things that look like leaks but would vanish under memory pressure.

**Comment by pcantrell** [comment]
> I notice you use a rather remarkable curly brace style that I've never seen before. Is this simply your preference, or wasn't it supposed to look like this?

It’s my personally preferred style, which reduces the non-information-bearing visual noise created by the ragged block boundaries of other styles. I came to it on my own back when I first learned C, but later found out it has a name (“Whitesmiths”). Apparently some study or other found it easier to read. Who knows? I like it. I don’t generally inflict on teams I work with, but like to use in my own code.
